x-pack/filebeat/input/salesforce/input.go

```diff
@@ -0,0 +1,506 @@
+func (s *salesforceInput) RunObject() error {
+	s.log.Debugf("Scrape Objects every %s", s.srcConfig.EventMonitoringMethod.Object.Interval)
+
+	var cursor mapstr.M
+	if !(s.cursor.Object.FirstEventTime == "" && s.cursor.Object.LastEventTime == "") {
+		object := make(mapstr.M)
+		if s.cursor.Object.FirstEventTime != "" {
+			object.Put("first_event_time", s.cursor.Object.FirstEventTime)
+		}
+		if s.cursor.Object.LastEventTime != "" {
+			object.Put("last_event_time", s.cursor.Object.LastEventTime)
+		}
+		cursor = mapstr.M{"object": object}
+	}
+	query, err := s.FormQueryWithCursor(s.config.EventMonitoringMethod.Object.Query, cursor)
+	if err != nil {
+		return fmt.Errorf("error forming based on cursor: %w", err)
+	}
+
+	res, err := s.soqlr.Query(query, false)
+	if err != nil {
+		return err
+	}
+
+	totalEvents := 0
+	firstEvent := true
+
+	for res.TotalSize() > 0 {
+		for _, rec := range res.Records() {
+			val := rec.Record().Fields()
+
+			jsonStrEvent, err := json.Marshal(val)
+			if err != nil {
+				return err
+			}
+
+			if firstEvent {
+				if timstamp, ok := val[s.config.EventMonitoringMethod.Object.Cursor.Field].(string); ok {
+					s.cursor.Object.FirstEventTime = timstamp
+				}
+			}
+
+			if timstamp, ok := val[s.config.EventMonitoringMethod.Object.Cursor.Field].(string); ok {
+				s.cursor.Object.LastEventTime = timstamp
+			}
+
+			err = publishEvent(s.publisher, s.cursor, jsonStrEvent, "Object")
+			if err != nil {
+				return err
+			}
+			firstEvent = false
+			totalEvents++
+		}
+
+		if res.MoreRecords() { // returns true if there are more records.
+			res, err = res.Next()
+			if err != nil {
+				return err
+			}
+		} else {
+			break
+		}
+	}
+	s.log.Debugf("Total events: %d", totalEvents)
+
+	return nil
+}
+
+// RunEventLogFile runs the EventLogFile method of the Event Monitoring API to
+// collect events.
+func (s *salesforceInput) RunEventLogFile() error {
+	s.log.Debugf("Scrape EventLogFiles every %s", s.srcConfig.EventMonitoringMethod.EventLogFile.Interval)
+
+	var cursor mapstr.M
+	if !(s.cursor.EventLogFile.FirstEventTime == "" && s.cursor.EventLogFile.LastEventTime == "") {
+		eventLogFile := make(mapstr.M)
+		if s.cursor.EventLogFile.FirstEventTime != "" {
+			eventLogFile.Put("first_event_time", s.cursor.EventLogFile.FirstEventTime)
+		}
+		if s.cursor.EventLogFile.LastEventTime != "" {
+			eventLogFile.Put("last_event_time", s.cursor.EventLogFile.LastEventTime)
+		}
+		cursor = mapstr.M{"event_log_file": eventLogFile}
+	}
+
+	query, err := s.FormQueryWithCursor(s.config.EventMonitoringMethod.EventLogFile.Query, cursor)
+	if err != nil {
+		return fmt.Errorf("error forming based on cursor: %w", err)
+	}
+
+	res, err := s.soqlr.Query(query, false)
+	if err != nil {
+		return err
+	}
+
+	if s.sfdcConfig.Client == nil {
+		return errors.New("internal error: salesforce configuration is not set properly")
+	}
+
+	totalEvents, firstEvent := 0, true
+	for res.TotalSize() > 0 {
```

Comment:
there's lots of repetition between the two event monitoring methods .. it would be nice if we could abstract that out. not a blocker for this PR, but a potential improvement later.

Reply:
I agree. But this `for` loop might look similar but the internals are quite different. But I identified some other parts in input.go that are duplicated using [dupl](https://github.com/mibk/dupl).

Like this chunk:

```
var cursor mapstr.M
	if !(s.cursor.Object.FirstEventTime == "" && s.cursor.Object.LastEventTime == "") {
		object := make(mapstr.M)
		if s.cursor.Object.FirstEventTime != "" {
			object.Put("first_event_time", s.cursor.Object.FirstEventTime)
		}
		if s.cursor.Object.LastEventTime != "" {
			object.Put("last_event_time", s.cursor.Object.LastEventTime)
		}
		cursor = mapstr.M{"object": object}
	}
```

But other duplicated parts are really small.